common/types/string.go

```diff
@@ -221,3 +240,447 @@ func stringStartsWith(s String, pre ref.Val) ref.Val {
+	return stringFormatWithLocale(s, args, String(defaultLocale))
+}
+
+func stringFormatWithLocale(s String, args ref.Val, localeVal ref.Val) ref.Val {
+	formatStr := string(s)
+	argList, ok := args.(traits.Lister)
+	if !ok {
+		return MaybeNoSuchOverloadErr(argList)
+	}
+	localeStrVal, ok := localeVal.(String)
+	if !ok {
+		return MaybeNoSuchOverloadErr(localeStrVal)
+	}
+	locale := localeStrVal.Value().(string)
+
+	i := 0
+	argIndex := 0
+	var builtStr strings.Builder
+	for i < len(formatStr) {
+		if formatStr[i] == '%' {
+			if i+1 < len(formatStr) && formatStr[i+1] == '%' {
+				err := builtStr.WriteByte('%')
+				if err != nil {
+					return NewErr("error writing format string: %s", err.Error())
+				}
+				i += 2
+				continue
+			} else {
+				newStrIndex, val, refErr := parseFormatArg(i, argIndex, formatStr, argList, locale)
+				if refErr != nil {
+					return refErr
+				}
+				_, err := builtStr.WriteString(val)
+				if err != nil {
+					return NewErr("error writing format string: %s", err.Error())
+				}
+				i = newStrIndex
+				argIndex++
+			}
+		} else {
+			err := builtStr.WriteByte(formatStr[i])
+			if err != nil {
+				return NewErr("error writing format string: %s", err.Error())
+			}
+			i++
+		}
+	}
+	return String(builtStr.String())
+}
+
+func parseFormatArg(lastStrIndex, argIndex int, formatStr string, args traits.Lister, locale string) (int, string, ref.Val) {
+	if lastStrIndex+1 >= len(formatStr) {
+		return -1, "", NewErr("unexpected end of string")
+	}
+	refSize := args.Size()
+	argSize, ok := refSize.(Int)
+	if !ok {
+		return -1, "", NewErr("couldn't convert arg size")
+	}
+	if Int(argIndex) >= argSize {
+		return -1, "", NewErr("index %d out of range", argIndex)
+	}
+
+	i := lastStrIndex
+	i, formatter, err := parseFormattingClause(i, formatStr)
+	if err != nil {
+		return -1, "", NewErr("could not parse formatting clause: %s", err)
+	}
+
+	var argStr string
+	argAny := args.Get(Int(argIndex))
+	if formatter == nil {
+		return -1, "", NewErr("formatter required")
+	} else {
+		argStr, err = formatter(argAny, locale)
+		if err != nil {
+			return -1, "", NewErr("error during formatting: %s", err)
+		}
+	}
+	return i, argStr, nil
+}
+
+type clauseImpl func(ref.Val, string) (string, error)
+
+func formatString(lastStrIndex int, formatStr string) (int, clauseImpl, error) {
+	return lastStrIndex + 1, func(arg ref.Val, locale string) (string, error) {
+		argStrVal := arg.ConvertToType(StringType)
+		argStr, ok := argStrVal.Value().(string)
+		if !ok {
+			return "", fmt.Errorf("could not convert argument %q to string", argStrVal)
+		}
+		return argStr, nil
+	}, nil
+}
+
+func formatFixed(lastStrIndex int, formatStr string, precision *int) (int, clauseImpl, error) {
+	if precision == nil {
+		return -1, nil, errors.New("precision required for fixed-point output")
+	}
+	return lastStrIndex + 1, func(arg ref.Val, locale string) (string, error) {
+		argFloatVal := arg.ConvertToType(DoubleType)
```

Comment:
Does this mean that we can pass in any type that converts to the target type of the formatter? Is that what we want?

We do need some way to CEL's timestamp and duration types (BTW we should add tests for these). So _maybe_ this is the right way to format those types? But right now, if you "%h" format a "123" string, won't this result in the the hex value of int(123) when what most people would expect is the hex values of the individual chars of the string?

Reply:
I think the current behavior might be a little too forgiving, the more I think about it. I'll make `%s` only work for specific types (and not numbers) + add tests to make sure those guardrails work, and the same for `%e` and `%f` (I'll hold off on the `%h` one since we were still working out the behavior there). 

I think `%s` for timestamp and duration would work well, I'll add a test in for those.

Reply:
(marking as resolved since `%s` is used for timestamp/duration now)